### PR TITLE
[3 of 8: Service Updates & Upgrades] Introduce ServicePlanBlocks

### PR DESCRIPTION
**Pull request by jfurrow**

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @MatApple

File: src/js/components/ServicePlanBlocks.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
+import classNames from 'classnames';
+import {Tooltip} from 'reactjs-components';
```
Comment (kennyt) [inline src/js/components/ServicePlanBlocks.js]: :abc:

File: src/js/components/ServicePlanBlocks.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  handleBlockMouseLeave() {
+    this.setState({hoveredBlock: null});
+  }
+
+  getUpgradeBlocks(phaseBlocks, activePhase) {
+    let blocks = phaseBlocks.getItems().map((block, blockIndex) => {
+      let {selectedBlock} = this.state;
+      block = new ServicePlanBlock(block);
```
Comment (kennyt) [inline src/js/components/ServicePlanBlocks.js]: we should not need to do this. why isn't the block already instantiated?

File: src/js/components/ServicePlanBlocks.js (hunk 1 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Done!

File: src/js/components/ServicePlanBlocks.js (hunk 2 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: It's not an instance of `Block`, though I agree it should be. Let me see what's up in the corresponding struct.

Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Alright, done!

File: src/js/components/ServicePlanBlocks.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      this.setState({selectedBlock: null});
+    }
+  }
+
+  handleBlockRestart(blockID, phaseID) {
+    ServicePlanStore.restartBlock(this.props.service.id, blockID, phaseID);
+  }
+
+  handleBlockForceComplete(blockID, phaseID) {
+    ServicePlanStore.forceCompleteBlock(this.props.service.id, blockID,
```
Comment (kennyt) [inline src/js/components/ServicePlanBlocks.js]: we usually do it 
```js
.forceCompleteBlock(
  this.props.service.id,
  blockID,
  phaseID
)
```

File: src/js/components/ServicePlanBlocks.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+      let isComplete = block.isComplete();
+
+      if (isComplete && hasDecisionPoint) {
+        hasDecisionPoint = false;
+      }
+
+      let blockClassName = classNames('service-plan-modal-details-block', {
+        'has-decision-point': hasDecisionPoint,
+        'is-active': isActive,
+        'is-selected': selectedBlock && selectedBlock.getID() === block.getID(),
```
Comment (kennyt) [inline src/js/components/ServicePlanBlocks.js]: `block.getID()` is used twice in this function. let's set it

File: src/js/components/ServicePlanBlocks.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
+import classNames from 'classnames';
```
Comment (kennyt) [inline src/js/components/ServicePlanBlocks.js]: you have a struct named `ServicePlanBlocks` and this component is named `ServicePlanBlocks`. can we try to name it something different?

File: src/js/components/ServicePlanBlocks.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import React from 'react';
+import {Tooltip} from 'reactjs-components';
+
+import IconUpgradeBlock from './icons/IconUpgradeBlock';
+import IconUpgradeBlockDecisionPoint from './icons/IconUpgradeBlockDecisionPoint';
+import ServicePlanStore from '../stores/ServicePlanStore';
+
+const METHODS_TO_BIND = [
+  'handleExternalClick',
+  'handleBlockMouseEnter',
```
Comment (daiyiwhale) [inline src/js/components/ServicePlanBlocks.js]: 🔡

File: src/js/components/ServicePlanBlocks.js (hunk 5 above)
Comment (daiyiwhale) [inline src/js/components/ServicePlanBlocks.js]: they are literally service plan blocks :'D It's an accurate name

File: src/js/components/ServicePlanBlocks.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    let {activePhase, block, blockIndex} = blockDetails;
+    let blockOptions = [];
+    let blockLabelSeparator = ': ';
+
+    if (block.isComplete() || block.isInProgress()) {
+      blockOptions.push(
+        <a
+          className="clickable"
+          key="restart"
+          onClick={this.handleBlockRestart.bind(this, block.getID(),
```
Comment (daiyiwhale) [inline src/js/components/ServicePlanBlocks.js]: you got double `block.getID()`s and `activePhase.getID()`s here too if you wanted to nip that (':

File: src/js/components/ServicePlanBlocks.js (hunk 3 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Done!

File: src/js/components/ServicePlanBlocks.js (hunk 4 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Done!

File: src/js/components/ServicePlanBlocks.js (hunk 5 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Doen!

File: src/js/components/ServicePlanBlocks.js (hunk 6 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Done!

File: src/js/components/ServicePlanBlocks.js (hunk 7 above)
Comment (jfurrow) [inline src/js/components/ServicePlanBlocks.js]: Done!

**Comment by daiyiwhale** [comment]
Looks great! (:

File: src/js/components/ServicePlanBlockDetails.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+  render() {
+    let {servicePlan} = this.props;
+    let activePhase = servicePlan.getPhases().getActive();
+    let phaseBlocks = activePhase.getBlocks();
+    let blockCount = phaseBlocks.getItems().length;
+
+    let blocksHeadingContent = `Completed ${phaseBlocks.getComplete().length}
+      of ${blockCount}`;
+
+    if (!!this.state.hoveredBlock || !!this.state.selectedBlock) {
```
Comment (kennyt) [inline src/js/components/ServicePlanBlockDetails.js]: let's store these state properties into variables. you can simple do `let {hoveredBlock, selectedBlock} = this.state` on top

Comment (jfurrow) [inline src/js/components/ServicePlanBlockDetails.js]: Done!

**Pull request merged**